### ISSUE TITLE
CCodec_BC6H::~CCodec_BC6H crashes in std::thread 

### ISSUE THREAD

**Issue by mckelley3v** [GPUOpen-Tools/Compressonator#48]
Microsoft Visual Studio Professional 2017
Version 15.3.5

```
   CMP_CompressOptions compressOptions = {}; // zero initialize struct
   compressOptions.dwSize = sizeof(compressOptions);
   compressOptions.nGPUDecode = GPUDecode_DIRECTX;
   compressOptions.nComputeWith = Compute_DIRECTX;
   compressOptions.nCompressionSpeed = CMP_Speed_Normal;
   compressOptions.bUseAdaptiveWeighting = TRUE;
   compressOptions.fWeightingRed = 0.2126;
   compressOptions.fWeightingGreen = 0.7152;
   compressOptions.fWeightingBlue = 0.0722f;
   compressOptions.bDXT1UseAlpha = isBC1 && hasAlpha;
   compressOptions.nAlphaThreshold = 128;
   compressOptions.dwnumThreads = 8; // crashes without this line

```

if CCodec_BC6H::m_Use_MultiThreading ends up being false, then the following loop in the destructor of CCodec_BC6H triggers std::terminate inside std::thread because the code is attempting to move a std::thread that hasn't yet completed its func.

```
        for (unsigned int i = 0; i < m_LiveThreads; i++)
        {
            std::thread& curThread = m_EncodingThreadHandle[i];

            curThread = std::thread(); // FAILS HERE BECAUSE curThread HASN'T TERMINATED
        }

```
I suspect the code that sets m_EncodeParameterStorage[i].exit = TRUE; and curThread.join() still needs to happen even if only one thread is alive.

**Comment by NPCompress** [comment]
@mckelley3v  Thanks for the detailed issue information. 

Notice that you have set   
 compressOptions.nComputeWith = Compute_DIRECTX;

This feature (Compute based compressed ) is not fully implemented in v3.0 and is currently been worked on!

We are looking into the case when compressOptions.dwnumThreads = 0 or 1, this should set m_Use_MultiThreading = false in the codec_BC6H code to turn off Multithreading

**Comment by xooiamd** [comment]
hi @mckelley3v  you are right, please try the destructor below. Let us know if the issue still exist with the following destructor. thanks! : 
CCodec_BC6H::~CCodec_BC6H()
{
    if (m_LibraryInitialized)
    {
        // Tell all the live threads that they can exit when they have finished any current work
        for (int i = 0; i < m_LiveThreads; i++)
        {
            // If a thread is in the running state then we need to wait for it to finish
            // any queued work from the producer before we can tell it to exit.
            //
            // If we don't wait then there is a race condition here where we have
            // told the thread to run but it hasn't yet been scheduled - if we set
            // the exit flag before it runs then its block will not be processed.
#pragma warning(push)
#pragma warning(disable : 4127)  //warning C4127: conditional expression is constant
            while (1)
            {
                if (m_EncodeParameterStorage == NULL)
                    break;
                if (m_EncodeParameterStorage[i].run != TRUE)
                {
                    break;
                }
            }
#pragma warning(pop)
            // Signal to the thread that it can exit
            m_EncodeParameterStorage[i].exit = TRUE;
        }

        for (DWORD dwThread = 0; dwThread < m_LiveThreads; dwThread++)
        {
            std::thread& curThread = m_EncodingThreadHandle[dwThread];

            curThread.join();
        }
           
        delete[] m_EncodingThreadHandle;
        m_EncodingThreadHandle = NULL;

        if (m_EncodeParameterStorage)
            delete[] m_EncodeParameterStorage;
        m_EncodeParameterStorage = NULL;

        for (int i = 0; i < m_NumEncodingThreads; i++)
        {
            if (m_encoder[i])
            {
                delete m_encoder[i];
                m_encoder[i] = NULL;
            }
        }

        if (m_decoder)
        {
            delete m_decoder;
            m_decoder = NULL;
        }

        m_LibraryInitialized = false;
    }
}
